File: LeetCode/DataStructure_II/Day16/199_binary_tree_right_side_view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None: return 
        queue = deque([root])
        candidate = deque()
        result = []
        while queue:
            curr = queue.popleft()
            if curr.left:
                candidate.append(curr.left)
            if curr.right: 
                candidate.append(curr.right)
            if len(queue) == 0:
                result.append(curr.val)
                while candidate: 
                    queue.append(candidate.popleft())
                    
            
        return result
            
        

File: LeetCode/DayChallenges/DayChallenge_2022_July/199_Binary_Tree_Right_Side_View.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 var rightSideView = function(root) {
  if (root === null) return [];
  const result = new Array(root.length);
  inorder(root, 0, result);
  return result;
};

const inorder = function(node, depth, arr) {
  if(node.left) inorder(node.left, depth + 1, arr);
  arr[depth] = node.val;
  if(node.right) inorder(node.right, depth + 1, arr);
}